snowflake/data_access_sync_to_target.go

```diff
@@ -0,0 +1,1659 @@
+			return uniqueMaskName, err
+		}
+	}
+
+	return uniqueMaskName, nil
+}
+
+func (s *AccessSyncer) removeMask(_ context.Context, maskName string, repo dataAccessRepository) error {
+	logger.Info(fmt.Sprintf("Remove mask %q", maskName))
+
+	existingPolicies, err := repo.GetPoliciesLike("MASKING", fmt.Sprintf("%s%s", maskName, "%"))
+	if err != nil {
+		return err
+	}
+
+	for _, policy := range existingPolicies {
+		err = repo.DropMaskingPolicy(policy.DatabaseName, policy.SchemaName, maskName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *AccessSyncer) createGrantsForTableOrView(doType string, permissions []string, fullName string, metaData map[string]map[string]struct{}) ([]Grant, error) {
+	// TODO: this does not work for Raito full names
+	sfObject := common.ParseFullName(fullName)
+	if sfObject.Database == nil || sfObject.Schema == nil || sfObject.Table == nil {
+		return nil, fmt.Errorf("expected fullName %q to have 3 parts (database.schema.view)", fullName)
+	}
+
+	grants := make([]Grant, 0, len(permissions)+2)
+
+	for _, p := range permissions {
+		if _, f := metaData[doType][strings.ToUpper(p)]; f {
+			grants = append(grants, Grant{p, doType, common.FormatQuery(`%s.%s.%s`, *sfObject.Database, *sfObject.Schema, *sfObject.Table)})
+		} else {
+			logger.Warn(fmt.Sprintf("Permission %q does not apply to type %s", p, strings.ToUpper(doType)))
+		}
+	}
+
+	if len(grants) > 0 {
+		grants = append(grants,
+			Grant{"USAGE", ds.Database, common.FormatQuery(`%s`, *sfObject.Database)},
+			Grant{"USAGE", ds.Schema, common.FormatQuery(`%s.%s`, *sfObject.Database, *sfObject.Schema)})
+	}
+
+	return grants, nil
+}
+
+func (s *AccessSyncer) getTablesForSchema(repo dataAccessRepository, database, schema string) ([]TableEntity, error) {
+	cacheKey := database + "." + schema
+
+	if tables, f := s.tablesPerSchemaCache[cacheKey]; f {
+		return tables, nil
+	}
+
+	tables := make([]TableEntity, 10)
+
+	err := repo.GetTablesInDatabase(database, schema, func(entity interface{}) error {
+		table := entity.(*TableEntity)
+		tables = append(tables, *table)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	s.tablesPerSchemaCache[cacheKey] = tables
+
+	return tables, nil
+}
+
+func (s *AccessSyncer) getSchemasForDatabase(repo dataAccessRepository, database string) ([]SchemaEntity, error) {
+	if schemas, f := s.schemasPerDataBaseCache[database]; f {
+		return schemas, nil
+	}
+
+	schemas := make([]SchemaEntity, 10)
+
+	err := repo.GetSchemasInDatabase(database, func(entity interface{}) error {
+		schema := entity.(*SchemaEntity)
+		schemas = append(schemas, *schema)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	s.schemasPerDataBaseCache[database] = schemas
+
+	return schemas, nil
+}
+
+func (s *AccessSyncer) getDatabases(repo dataAccessRepository) ([]DbEntity, error) {
+	if s.databasesCache != nil {
+		return s.databasesCache, nil
+	}
+
+	var err error
+	s.databasesCache, err = repo.GetDatabases()
```

Comment:
Just wondering, should we take exclude-database items into account?

If not, should be do an extra filtering for the databases we are going to loop over for getting the database-roles based on that parameter?

Reply:
We should definitely only look at the database we are importing, yes. So just like for the data-source part, we need to filter out the right databases here.